Mathlib/Tactic/SolveByElim.lean

```diff
@@ -15,40 +18,301 @@ We'll gradually bring it up to feature parity.
+
+/-- Create or modify a `Config` which calls `intro` on each goal before applying lemmas. -/
+def intros (cfg : Config := {}) : Config :=
+  mainGoalProc cfg fun g => do pure [(← g.intro1P).2]
+
+/--
+Create or modify a `Config` which rejects branches for which `test`,
+applied to the instantiations of the original goals, fails or returns `false`.
+-/
+def testPartialSolutions (cfg : Config := {}) (test : List Expr → MetaM Bool) : Config :=
+{ cfg with
+  proc := fun orig goals => do
+    let .true ← test (← orig.mapM fun m => m.withContext do instantiateMVars (.mvar m)) | failure
+    cfg.proc orig goals }
+
+/--
+Create or modify a `Config` which rejects complete solutions for which `test`,
+applied to the instantiations of the original goals, fails or returns `false`.
+-/
+def testSolutions (cfg : Config := {}) (test : List Expr → MetaM Bool) : Config :=
+  cfg.testPartialSolutions fun sols => do
+    trace[Meta.Tactic.solveByElim] sols
+    if sols.any Expr.hasMVar then
+      pure true
+    else
+      test sols
+
+/--
+Create or modify a `Config` which only accept solutions
+for which every expression in `use` appears as a subexpression.
+-/
+def requireUsingAll (cfg : Config := {}) (use : List Expr) : Config :=
+  cfg.testSolutions fun sols => do
+    trace[Meta.Tactic.solveByElim] sols
+    trace[Meta.Tactic.solveByElim] m!"{use.map fun e => sols.map fun s => e.occurs s}"
+    pure <| use.all fun e => sols.any fun s => e.occurs s
+
+end Config
+
+/--
+Elaborate a list of lemmas and local context.
+See `mkAssumptionSet` for an explanation of why this is needed.
+-/
+def elabContextLemmas (g : MVarId) (lemmas : List (TermElabM Expr)) (ctx : TermElabM (List Expr)) :
+    MetaM (List Expr) :=
+  g.withContext (Elab.Term.TermElabM.run' do pure ((← lemmas.mapM id) ++ (← ctx)))
+
+/--
+Solve a collection of goals by repeatedly applying lemmas, backtracking as necessary.
+
+Arguments:
+* `cfg : Config` additional configuration options
+  (options for `apply`, maximum depth, and custom flow control)
+* `lemmas : List (TermElabM Expr)` lemmas to apply.
+  These are thunks in `TermElabM` to avoid stuck metavariables.
+* `ctx : TermElabM (List Expr)` monadic function returning the local hypotheses to use.
+* `goals : List MVarId` the initial list of goals for `solveByElim`
+
+Returns a list of suspended goals, if it succeeded on all other subgoals.
+By default `cfg.suspend` is `false,` `cfg.discharge` fails, and `cfg.failAtMaxDepth` is `true`,
+and so the returned list is always empty.
+Custom wrappers (e.g. `apply_assumption` and `apply_rules`) may modify this behaviour.
+-/
+def solveByElim (cfg : Config) (lemmas : List (TermElabM Expr)) (ctx : TermElabM (List Expr))
+    (goals : List MVarId) : MetaM (List MVarId) := do
+  -- We handle `cfg.symm` by saturating hypotheses of all goals using `symm`.
+  let goals ← if cfg.symm then
+    goals.mapM fun g => g.symmSaturate
+  else
+    pure goals
+
+  match cfg.exfalso, goals with
```

Comment:
Just noting: this approach to `exfalso` and `symm` is significantly different from [how mathlib3 does it](https://github.com/leanprover-community/mathlib/blob/master/src/tactic/core.lean#L1103-L1107) (i.e. by trying to apply `exfalso` and `symmetry` at every step long the way). It's not obvious to me whether its better or worse, but we should keep this in mind as a possible source of performance discrepancy.

Reply:
I should have written an explanation here.

I first implemented the mathlib3 approach, and realised it was really slow (adds about 50% to the library search test file). I then switched to this approach, which has an almost undetectable slow down. 

I'll document this soon.